explorer/db/hooks_test.go

```diff
@@ -0,0 +1,153 @@
+	"github.com/omni-network/omni/explorer/db/ent"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+type results struct {
+	blocks   []ent.Block
+	messages []*ent.Msg
+	receipts []*ent.Receipt
+}
+
+type prerequisite func(t *testing.T, ctx context.Context) results
+
+func TestMsgHooks(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	type want struct {
+		BlockCount        int
+		MsgCount          int
+		ReceiptCount      int
+		StrayReceiptCount int
+		StrayMessageCount int
+	}
+
+	tests := []struct {
+		name          string
+		prerequisites prerequisite // These functions create entries on our db before the evaluation
+		wantErr       bool
+		want          want
+	}{
+		{
+			name: "create_initial_block_with_msg_then_following_contain_receipts",
+			prerequisites: func(t *testing.T, ctx context.Context) results {
+				t.Helper()
+				client := db.CreateTestEntClient(t)
+				t.Cleanup(func() {
+					if err := client.Close(); err != nil {
+						t.Error(err)
+					}
+				})
+				blocks := db.CreateTestBlocks(t, ctx, client, 2)
+
+				var messages []*ent.Msg
+				var receipts []*ent.Receipt
+				for _, b := range blocks {
+					messages = append(messages, b.QueryMsgs().AllX(ctx)...)
+					receipts = append(receipts, b.QueryReceipts().AllX(ctx)...)
+				}
+
+				return results{
+					blocks:   blocks,
+					messages: messages,
+					receipts: receipts,
+				}
+			},
+			want: want{
+				BlockCount:        2,
+				MsgCount:          2,
+				ReceiptCount:      1,
+				StrayReceiptCount: 0,
+				StrayMessageCount: 1,
+			},
+			wantErr: false,
+		},
+		{
+			name: "create_block_then_receipt_then_msg",
+			prerequisites: func(t *testing.T, ctx context.Context) results {
+				t.Helper()
+				client := db.CreateTestEntClient(t)
+				t.Cleanup(func() {
+					if err := client.Close(); err != nil {
+						t.Error(err)
+					}
+				})
+				destChainID := uint64(2)
+				streamOffset := uint64(0)
+				block1 := db.CreateTestBlock(t, ctx, client, 0)
+				block2 := db.CreateTestBlock(t, ctx, client, 1)
+				receipt := db.CreateReceipt(t, ctx, client, block1, destChainID, streamOffset)
+				msg := db.CreateXMsg(t, ctx, client, block2, destChainID, streamOffset)
+
+				messages := []*ent.Msg{msg}
+				receipts := []*ent.Receipt{receipt}
+				blocks := []ent.Block{block1, block2}
+
+				return results{
+					blocks:   blocks,
+					messages: messages,
+					receipts: receipts,
+				}
+			},
+			want: want{
+				BlockCount:        2,
+				MsgCount:          1,
+				ReceiptCount:      1,
+				StrayReceiptCount: 0,
+				StrayMessageCount: 0,
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+			results := tt.prerequisites(t, ctx)
+			actual := want{
+				len(results.blocks),
+				len(results.messages),
+				len(results.receipts),
+				calcStrayReceipts(ctx, results.receipts),
+				calcStrayMessages(ctx, results.messages),
+			}
+
+			if !cmp.Equal(tt.want, actual) {
```

Comment:
Maybe an overkill (feel free to ignore), but I would compare the entire `results` struct against a known value. This way you don't only make sure that there are no stray msgs/receipts but also you make sure that the each receipt is connected to the correct message as well.

Reply:
hmmmmm let me think about how to do this. I have to write another hook and subsequent test so I'll try to wrap those changes into that PR.